# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/base/DOMIntersectionObserver.cpp
# Commit: 1c5275a57680
# Full Hash: 1c5275a5768005c225df8a937d3e36d8789f8da2
# Author: Hiroyuki Ikezoe <hikezoe.birchill@mozilla.com>
# Date: 2020-03-05 09:55:41
# Description:
#   Bug 1619857 - Early return from GetOopIframeMetrics in cases where either the presshell or the docshell is being destroyed. r=emilio
#   
#   I gave up writing crash tests for this since it's quite hard to destroy
#   an OOP iframe during processing IntersectionObserver's update step in the
#   OOP process.
# ==============================================================================

diff -r 97a8f04641de -r 1c5275a57680 dom/base/DOMIntersectionObserver.cpp
--- a/dom/base/DOMIntersectionObserver.cpp	Wed Mar 04 09:14:44 2020 +0000
+++ b/dom/base/DOMIntersectionObserver.cpp	Thu Mar 05 01:56:48 2020 +0000
@@ -441,7 +441,7 @@
   MOZ_ASSERT(rootDoc && !rootDoc->IsTopLevelContentDocument());
 
   PresShell* rootPresShell = rootDoc->GetPresShell();
-  if (!rootPresShell) {
+  if (!rootPresShell || rootPresShell->IsDestroying()) {
     return Nothing();
   }
 
@@ -450,16 +450,18 @@
     return Nothing();
   }
 
+  BrowserChild* browserChild = BrowserChild::GetFrom(rootDoc->GetDocShell());
+  if (!browserChild) {
+    return Nothing();
+  }
+  MOZ_DIAGNOSTIC_ASSERT(!browserChild->IsTopLevel());
+
   nsRect inProcessRootRect;
   if (nsIScrollableFrame* scrollFrame =
           rootPresShell->GetRootScrollFrameAsScrollable()) {
     inProcessRootRect = scrollFrame->GetScrollPortRect();
   }
 
-  nsIDocShell* docShell = rootDoc->GetDocShell();
-  BrowserChild* browserChild = BrowserChild::GetFrom(docShell);
-  MOZ_ASSERT(browserChild && !browserChild->IsTopLevel());
-
   Maybe<LayoutDeviceRect> remoteDocumentVisibleRect =
       browserChild->GetTopLevelViewportVisibleRectInSelfCoords();
   if (!remoteDocumentVisibleRect) {
